File: ios/ClawChatKit/Sources/ClawChatKit/Protocol/ControlMessages.swift
import Foundation

// MARK: - Presence Status

public enum PresenceStatus: String, Codable, Sendable {
    case online, away, offline
}

// MARK: - Typing

public struct Typing: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let agentId: String?
    public let active: Bool
    public let label: String?

    public init(agentId: String? = "default", active: Bool) {
        self.type = "typing"
        self.id = UUID().uuidString
        self.ts = Int64(Date().timeIntervalSince1970 * 1000)
        self.agentId = agentId
        self.active = active
        self.label = nil
    }
}

// MARK: - Presence

public struct Presence: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let status: PresenceStatus?
    public let online: Bool?
    public let gatewayId: String?
}

// MARK: - Status Request / Response

public struct StatusRequest: BaseMessage, Sendable {
    public let type: String = "status.request"
    public let id: String
    public let ts: Int64

    public init() {
        self.id = UUID().uuidString
        self.ts = Int64(Date().timeIntervalSince1970 * 1000)
    }
}

public struct AgentMeta: Codable, Sendable {
    public let name: String?
    public let model: String?
    public let avatar: String?
}

public struct StatusResponse: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let gatewayOnline: Bool
    public let agents: [String]?
    public let agentsMeta: [String: AgentMeta]?
    public let connectedDevices: Int?
}

File: android/clawchat-kit/src/main/kotlin/com/clawchat/kit/protocol/ControlMessages.kt
package com.clawchat.kit.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

// MARK: - Presence Status

@Serializable
enum class PresenceStatus {
    @SerialName("online") ONLINE,
    @SerialName("away") AWAY,
    @SerialName("offline") OFFLINE,
}

// MARK: - Typing

@Serializable
data class Typing(
    override val type: String = "typing",
    override val id: String = UUID.randomUUID().toString(),
    override val ts: Long = System.currentTimeMillis(),
    val agentId: String? = "default",
    val active: Boolean,
    val label: String? = null,
) : BaseMessage

// MARK: - Presence

@Serializable
data class Presence(
    override val type: String,
    override val id: String,
    override val ts: Long,
    val status: PresenceStatus? = null,
    val online: Boolean? = null,
    val gatewayId: String? = null,
) : BaseMessage

// MARK: - Status Request

@Serializable
data class StatusRequest(
    override val type: String = "status.request",
    override val id: String = UUID.randomUUID().toString(),
    override val ts: Long = System.currentTimeMillis(),
) : BaseMessage

// MARK: - Status Response

@Serializable
data class StatusResponse(
    override val type: String,
    override val id: String,
    override val ts: Long,
    val gatewayOnline: Boolean,
    val agents: List<String>? = null,
    val connectedDevices: Int? = null,
) : BaseMessage
